### PR TITLE
[feature](IO) Replace file writer's finalize function with async close

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes

Issue Number: close #xxx

Define the semantics of closed() and close()

1. The semantics of close is to ensure data persistence to the file system (FS) and release resources (e.g., file descriptors or connections).
2. The semantics of closed indicates that the close function has been called, and regardless of whether close was successful or not, closed will be true. When closed is true, all append calls are rejected.
3. Regardless of whether close has been called or whether it was successful, the FileWriter destructor must ensure the release of resources (e.g., file descriptors or connections).
4. The usage pattern for close(nonblock) is: after calling close(nonblock), you need to call close(block) again to synchronize the close status and confirm whether close was successful or not.
5. If close is called multiple times, any situation other than 4. is unexpected and does not conform to expectations. Unexpected calls should directly return InternalError.

规定好closed()和close()的语义
1. close 语义为保证数据持久化到 FS，并释放资源（e.g. fd / connection）
2. closed 语义表示调用过 close 函数，无论 close 成功与否 closed 都是 true，在 closed == true 时拒绝一切 append 调用
3. 无论是否调用过 close，close 是否成功，FileWriter 析构后都要保证释放资源 （e.g. fd / connection） 
4. close(nonblock) 的使用姿势是：close(nonblock) 后，需要再调用一次 close(block) 来同步 close 的状态，确认 close 成功与否
5. 如果有多次调用 close 的情况，除 4. 外的情况是不符合预期的，不符合预期的调用直接返回 InternalError

img_v3_02aq_f6185458-e743-4f76-836c-a13c409532ag

```
     append
     .----.
     |    |
     |    v
    .--------. non-blocking  .--------.   blocking    .--------.
--->| opened | ------------> | async  | ------------> | closed |
    '--------'     close     | closed |     close     '--------'
         |                   '--------'                    ^
         |                                                 |
         '-------------------------------------------------'
                           blocking close
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

File: be/src/common/config.cpp [hunk 1]
```diff
@@ -1257,6 +1257,11 @@ DEFINE_Int64(num_s3_file_upload_thread_pool_max_thread, "64");
 // The max ratio for ttl cache's size
 DEFINE_mInt64(max_ttl_cache_ratio, "90");
 
+// The min thread num for NonBlockCloseThreadPool
+DEFINE_Int64(min_non_block_close_thread_num, "12");
```
Comment (gavinchou) [inline be/src/common/config.cpp:1261]: naming
min_nonblocking_close_thread_num
max_nonblocking_close_thread_num

File: be/src/io/fs/file_writer.h [hunk 2]
```diff
@@ -54,25 +55,26 @@ class FileWriter {
     FileWriter(const FileWriter&) = delete;
     const FileWriter& operator=(const FileWriter&) = delete;
 
-    // Normal close. Wait for all data to persist before returning.
-    // If there is no data appended, an empty file will be persisted.
-    virtual Status close() = 0;
+    Status close(bool non_block = false);
```
Comment (gavinchou) [inline be/src/io/fs/file_writer.h:58]: usually we dont use doubly "negative" to present default value,
better change the signature to 
```
    Status close(bool blocking = true);
```

Comment (gavinchou) [reply be/src/io/fs/file_writer.h:58]: also add sufficient comment for this interface, behavior and params

Comment (gavinchou) [reply be/src/io/fs/file_writer.h:58]: we may need UT for the async/nonblocking test

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Review by platoneko — APPROVED** [review]

**Pull request merged**